engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -380,6 +513,122 @@ private void handleCreateTemplateFromSnapshot(SnapshotInfo snapshotInfo, Templat
+                    errMsg = copyCmdAnswer.getDetails();
+                }
+                else {
+                    errMsg = "Unable to create volume from snapshot";
+                }
+            }
+        }
+        catch (Exception ex) {
+            errMsg = ex.getMessage() != null ? ex.getMessage() : "Copy operation failed in 'StorageSystemDataMotionStrategy.handleCreateVolumeFromSnapshotBothOnStorageSystem'";
+        } finally {
+
+            // detach and remove access after the volume is created
+            DiskTO disk = new DiskTO(volumeInfo.getTO(), volumeInfo.getDeviceId(), volumeInfo.getPath(), volumeInfo.getVolumeType());
+            DettachCommand cmd = new DettachCommand(disk, null);
+            long storagePoolId = volumeInfo.getPoolId();
+            StoragePoolVO storagePool = _storagePoolDao.findById(storagePoolId);
+
+            cmd.setManaged(true);
+            cmd.setStorageHost(storagePool.getHostAddress());
+            cmd.setStoragePort(storagePool.getPort());
+            cmd.set_iScsiName(volumeInfo.get_iScsiName());
+
+            try {
+                DettachAnswer dettachAnswer = (DettachAnswer) _agentMgr.send(hostVO.getId(), cmd);
+
+                if (!dettachAnswer.getResult()) {
+                    throw new CloudRuntimeException("Error detaching volume:" + dettachAnswer.getDetails());
+                }
+
+            } catch (Exception e) {
+                LOGGER.warn("Error detaching volume " + volumeInfo.getId() + " Error: " + e.getMessage());
+            }
+
+            _volumeService.revokeAccess(volumeInfo, hostVO, destDataStore);
+        }
+
+        CopyCommandResult result = new CopyCommandResult(null, copyCmdAnswer);
+
+        result.setResult(errMsg);
+
+        callback.complete(result);
+
+    }
+
+    private void performCleanupCacheStorage(DataObject destOnStore) {
+        destOnStore.processEvent(Event.DestroyRequested);
+
+        DeleteCommand deleteCommand = new DeleteCommand(destOnStore.getTO());
+        EndPoint ep = selector.select(destOnStore);
+        try {
+            if (ep == null) {
+                LOGGER.warn("Unable to cleanup staging NFS because no endpoint was found " +
+                "Object: " + destOnStore.getType() + " ID: " + destOnStore.getId());
+
+                destOnStore.processEvent(Event.OperationFailed);
+            } else {
+                Answer deleteAnswer = ep.sendMessage(deleteCommand);
+                if (deleteAnswer != null && deleteAnswer.getResult()) {
+                    LOGGER.warn("Unable to cleanup staging NFS " + deleteAnswer.getDetails() +
+                    "Object: " + destOnStore.getType() + " ID: " + destOnStore.getId());
+                    destOnStore.processEvent(Event.OperationFailed);
+                }
+            }
+
+            destOnStore.processEvent(Event.OperationSuccessed);
+        } catch (Exception e) {
```

Comment:
We don't want to error out when we fail to clean up cache storage. I have added a statemachine transition using the event `OperationFailed` in the catch block